exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/common/HashPartition.java

```diff
@@ -0,0 +1,579 @@
+    currHVVector = new IntVector(MaterializedField.create(HASH_VALUE_COLUMN_NAME, HVtype), allocator);
+    currHVVector.allocateNew(RECORDS_PER_BATCH);
+  }
+
+  /**
+   *  Spills if needed
+   */
+  public void appendInnerRow(VectorContainer buildContainer, int ind, int hashCode, HashJoinMemoryCalculator.BuildSidePartitioning calc) {
+
+    int pos = currentBatch.appendRow(buildContainer,ind);
+    currHVVector.getMutator().set(pos, hashCode);   // store the hash value in the new column
+    if ( pos + 1 == RECORDS_PER_BATCH ) {
+      boolean needsSpill = isSpilled || calc.shouldSpill();
+      completeAnInnerBatch(true, needsSpill);
+    }
+  }
+
+  /**
+   *  Outer always spills when batch is full
+   *
+   */
+  public void appendOuterRow(int hashCode, int recordsProcessed) {
+    int pos = currentBatch.appendRow(probeBatch.getContainer(),recordsProcessed);
+    currHVVector.getMutator().set(pos, hashCode);   // store the hash value in the new column
+    if ( pos + 1 == RECORDS_PER_BATCH ) {
+      completeAnOuterBatch(true);
+    }
+  }
+
+  public void completeAnOuterBatch(boolean toInitialize) {
+    completeABatch(toInitialize, true);
+  }
+  public void completeAnInnerBatch(boolean toInitialize, boolean needsSpill) {
+    completeABatch(toInitialize, needsSpill);
+  }
+  /**
+   *     A current batch is full (or no more rows incoming) - complete processing this batch
+   * I.e., add it to its partition's tmp list, if needed - spill that list, and if needed -
+   * (that is, more rows are coming) - initialize with a new current batch for that partition
+   * */
+  private void completeABatch(boolean toInitialize, boolean needsSpill) {
+    if ( currentBatch.hasRecordCount() && currentBatch.getRecordCount() > 0) {
+      currentBatch.add(currHVVector);
+      currentBatch.buildSchema(BatchSchema.SelectionVectorMode.NONE);
+      tmpBatchesList.add(currentBatch);
+      partitionBatchesCount++;
+
+      long batchSize = new RecordBatchSizer(currentBatch).actualSize();
+      inMemoryBatchStats.add(new HashJoinMemoryCalculator.BatchStat(currentBatch.getRecordCount(), batchSize));
+
+      partitionInMemorySize += batchSize;
+      numInMemoryRecords += currentBatch.getRecordCount();
+    } else {
+      freeCurrentBatchAndHVVector();
+    }
+    if ( needsSpill ) { // spill this batch/partition and free its memory
+      spillThisPartition();
+    }
+    if ( toInitialize ) { // allocate a new batch and HV vector
+      allocateNewCurrentBatchAndHV();
+    } else {
+      currentBatch = null;
+      currHVVector = null;
+    }
+  }
+
+  /**
+   *  Append the incoming batch (actually only the vectors of that batch) into the tmp list
+   */
+  public void appendBatch(VectorAccessible batch) {
+    assert numPartitions == 1;
+    int recordCount = batch.getRecordCount();
+    currHVVector = new IntVector(MaterializedField.create("Hash_Values", HVtype), allocator);
```

Comment:
Use already defined constant for Hash_Values.

Reply:
Done